R/simulate_linreg.R

```diff
@@ -126,106 +96,72 @@ compute_s2g <- function(RL, beta) {
+  n_snps = ncol(G)
+  ncausal = parse_num_causal_snps(ncausal)
+  n_causal = if(ncausal$is_pct){
     max(1, round(ncausal$value * n_snps))
-  } else {
+  }else{
     min(ncausal$value, n_snps)
   }
+  B = matrix(0, nrow =  ncol(G), ncol = ntrait)
   
-  if (eqtl_h2 != 0) {
-    c_qtls <- sample(seq_len(n_snps), n_qtls)
-    b_qtls <- numeric(n_snps)
-    b_qtls[c_qtls] <- rnorm(n_qtls, mean = 0, sd = sqrt(eqtl_h2 / n_qtls))
-    
-    if (rescale) {
-      s2g <- compute_s2g(RL, b_qtls)
-      b_qtls <- b_qtls * sqrt(eqtl_h2 / s2g)
+  if(shared_pattern == "all"){
+    if(is_h2g_total){
+      causal_index = sample(seq_len(n_snps), n_causal)
+      beta = numeric(n_snps)
+      beta[causal_index] = 1
+      for(i in 1:ntrait){
+        B[,i] = beta
+      }
+    }else{
+      causal_index = sample(seq_len(n_snps), n_causal)
+    beta = numeric(n_snps)
+    beta[causal_index[1]] = 1
+    var_vector = apply(as.matrix(G[,causal_index]), 2, var)
+    beta[causal_index] = sqrt(beta[causal_index[1]]^2 * var_vector[1] / var_vector)
+    for(i in 1:ntrait){
+        B[,i] = beta
+      }
     }
-  } else {
-    b_qtls <- numeric(n_snps)
-  }
-  
-  return(b_qtls)
-}
-
-#' Simulate Genotypes
-#'
-#' Sample genotypes from a multivariate normal approximation.
-#'
-#' @param RL Lower Cholesky factor of the p x p LD matrix for the population.
-#' @param n Number of genotypes to sample.
-#' @return Centered and scaled genotype matrix.
-#' @importFrom stats rnorm
-#' @export
-#'
-#' @examples
-#' R <- matrix(c(1, 0.5, 0.5, 1), nrow = 2)  # LD matrix
-#' n <- 100                                  # Number of genotypes to sample
-#' RL <- get_lower_chol(R)                              # Compute lower Cholesky decomposition
-#' G <- sim_geno(RL, n)                       # Simulate genotypes
-sim_geno <- function(RL, n) {
-  p <- nrow(RL)
-
-  G <- RL %*% matrix(rnorm(n * p), ncol = p)
-  G <- scale(G)  # Center and scale
-
-  return(G)
-}
-
-#' Simulate a Polygenic Trait
-#'
-#' Simulate a complex trait as a function of latent genetic values and environmental noise.
-#'
-#' @param g Vector of latent genetic values.
-#' @param h2g Heritability of the trait in the population.
-#' @return Simulated phenotype.
-#' @importFrom stats rnorm
-#' @export
-#'
-#' @examples
-#' R <- matrix(c(1, 0.5, 0.5, 1), nrow = 2)  # LD matrix
-#' n <- 100                                  # Number of genotypes to sample
-#' RL <- get_lower_chol(R)                              # Compute lower Cholesky decomposition
-#' G <- sim_geno(RL, n)                       # Simulate genotypes
-#' ncausal <- list(value = 50, is_pct = TRUE)  # 50% of observed SNPs
-#' eqtl_h2 <- 0.5                              # Heritability of gene expression
-#' b <- sim_beta(RL, ncausal, eqtl_h2) 
-#' g <- G %*% b                   # latent genetic values
-#' y <- simulate_polygenic_trait(g, eqtl_h2)  # Simulate the complex trait
-simulate_polygenic_trait <- function(g, h2g) {
```

Comment:
Any reason to delete this function? It is still relevant if we need to do TWAS related simulations adding polygenic background to data. Please bring it back even if you don't use it --- unless you have a good reason to delete.